module/icp/algs/sha2/sha2_generic.c

```diff
@@ -0,0 +1,574 @@
+		total += (64 - pos) * 8;
+		data += 64 - pos;
+		pos = 0;
+	}
+
+	if (len >= 64) {
+		uint32_t blocks = len / 64;
+		uint32_t bytes = blocks * 64;
+		kfpu_begin();
+		ops->transform(ctx->state, data, blocks);
+		kfpu_end();
+		len -= bytes;
+		total += (bytes) * 8;
+		data += bytes;
+	}
+	memcpy(m + pos, data, len);
+
+	pos += len;
+	total += len * 8;
+	ctx->count[0] = pos;
+	ctx->count[1] = total;
+}
+
+static void sha512_update(sha512_ctx *ctx, const uint8_t *data, size_t len)
+{
+	uint64_t pos = ctx->count[0];
+	uint64_t total = ctx->count[1];
+	uint8_t *m = ctx->wbuf;
+	const sha512_ops_t *ops = ctx->ops;
+
+	if (pos && pos + len >= 128) {
+		memcpy(m + pos, data, (size_t)(128 - pos));
+		kfpu_begin();
+		ops->transform(ctx->state, m, 1);
+		kfpu_end();
+		len -= 128 - pos;
+		total += (128 - pos) * 8;
+		data += 128 - pos;
+		pos = 0;
+	}
+
+	if (len >= 128) {
+		uint64_t blocks = len / 128;
+		uint64_t bytes = blocks * 128;
+		kfpu_begin();
+		ops->transform(ctx->state, data, blocks);
+		kfpu_end();
+		len -= bytes;
+		total += (bytes) * 8;
+		data += bytes;
+	}
+	memcpy(m + pos, data, (size_t)(len));
+
+	pos += len;
+	total += len * 8;
+	ctx->count[0] = pos;
+	ctx->count[1] = total;
+}
+
+static void sha256_final(sha256_ctx *ctx, uint8_t *result, int bits)
+{
+	uint64_t mlen, pos = ctx->count[0];
+	uint8_t *m = ctx->wbuf;
+	uint32_t *R = (uint32_t *)result;
+	const sha256_ops_t *ops = ctx->ops;
+
+	m[pos++] = 0x80;
+	if (pos > 56) {
+		memset(m + pos, 0, (size_t)(64 - pos));
+		kfpu_begin();
+		ops->transform(ctx->state, m, 1);
+		kfpu_end();
+		pos = 0;
+	}
+
+	memset(m + pos, 0, (size_t)(56 - pos));
+	mlen = cpu_to_be64(ctx->count[1]);
+	memcpy(m + (64 - 8), &mlen, 64 / 8);
+	kfpu_begin();
+	ops->transform(ctx->state, m, 1);
+	kfpu_end();
+
+	switch (bits) {
+	case 224: /* 28 */
+		R[0] = cpu_to_be32(ctx->state[0]);
+		R[1] = cpu_to_be32(ctx->state[1]);
+		R[2] = cpu_to_be32(ctx->state[2]);
+		R[3] = cpu_to_be32(ctx->state[3]);
+		R[4] = cpu_to_be32(ctx->state[4]);
+		R[5] = cpu_to_be32(ctx->state[5]);
+		R[6] = cpu_to_be32(ctx->state[6]);
+		break;
+	case 256: /* 32 */
+		R[0] = cpu_to_be32(ctx->state[0]);
+		R[1] = cpu_to_be32(ctx->state[1]);
+		R[2] = cpu_to_be32(ctx->state[2]);
+		R[3] = cpu_to_be32(ctx->state[3]);
+		R[4] = cpu_to_be32(ctx->state[4]);
+		R[5] = cpu_to_be32(ctx->state[5]);
+		R[6] = cpu_to_be32(ctx->state[6]);
+		R[7] = cpu_to_be32(ctx->state[7]);
+		break;
+	}
+
+	memset(ctx, 0, sizeof (*ctx));
+}
+
+static void sha512_final(sha512_ctx *ctx, uint8_t *result, int bits)
+{
+	uint64_t mlen, pos = ctx->count[0];
+	uint8_t *m = ctx->wbuf, *r;
+	uint64_t *R = (uint64_t *)result;
+	const sha512_ops_t *ops = ctx->ops;
+
+	m[pos++] = 0x80;
+	if (pos > 112) {
+		memset(m + pos, 0, (size_t)(128 - pos));
+		kfpu_begin();
+		ops->transform(ctx->state, m, 1);
+		kfpu_end();
+		pos = 0;
+	}
+
+	memset(m + pos, 0, (size_t)(128 - pos));
+	mlen = cpu_to_be64(ctx->count[1]);
+	memcpy(m + (128 - 8), &mlen, 64 / 8);
+	kfpu_end();
```

Comment:
Scratch that, reverse it.

Reply:
Hello @rincebrain - this branch is just an RFC ... please do not test in deep, it's full of bugs, I need to find in the weekend ;-)
But the [EdonR shrinking](https://github.com/openzfs/zfs/pull/13618) and the [BLAKE3 tuneable fix](https://github.com/openzfs/zfs/pull/13725) are ready for a review.

The kfpu_begin()/end() are on my TODO for this branch here.

Reply:
I didn't spend a bunch of time testing it, I just assumed it would at least run enough to try.

I'd advise marking the PR as a draft, and putting a more explicit disclaimer about it being known to not function correctly, if that's the case, otherwise some people who test open PRs if they seem interesting might try it and be distressed.

And the specific thing I was pointing out was that you had reversed `kfpu_end` and `kfpu_begin` in that one case, not that you were missing them more generally or should delete them.

I was waiting for there to be a few days when those two PRs didn't get an update to review them, as force-pushing over the entire branch makes it difficult to tell what's changed, and the last two times I started reviewing one of them, they got updated before I finished.

Reply:
For the generic methods the kfpu begin/end was not okay there... I will put it into the transform() function itself.

Reply:
> I was waiting for there to be a few days when those two PRs didn't get an update to review them, as force-pushing over the entire branch makes it difficult to tell what's changed, and the last two times I started reviewing one of them, they got updated before I finished.

I am sorry for this. I rebased the [Edon-R](https://github.com/openzfs/zfs/pull/13618) branch because some reference for this https://github.com/openzfs/zfs/pull/13710 was requested by @gmelikov 

And the next update to the sha2-rework will be a seperate commit - okay :)